firebase-common/src/main/java/com/google/firebase/emulators/EmulatorSettings.java

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.emulators;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.google.android.gms.common.annotation.KeepForSdk;
+import com.google.firebase.components.Preconditions;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Settings that control which Firebase services should access a local emulator, rather than
+ * production.
+ *
+ * <p>TODO(samstern): Un-hide this once Firestore, Database, and Functions are implemented
+ *
+ * @see com.google.firebase.FirebaseApp#enableEmulators(EmulatorSettings)
+ * @hide
+ */
+@KeepForSdk
+public class EmulatorSettings {
+
+  public static final class Builder {
+
+    private final Map<FirebaseEmulators, EmulatedServiceSettings> settingsMap = new HashMap<>();
+
+    /** Constructs an empty builder. */
+    public Builder() {}
+
+    /**
+     * Specify the emulator settings for a single service.
+     *
+     * @param emulator the emulated service.
+     * @param settings the emulator settings.
+     * @return the builder, for chaining.
+     */
+    @NonNull
+    public Builder addEmulatedService(
+        @NonNull FirebaseEmulators emulator, @NonNull EmulatedServiceSettings settings) {
+      Preconditions.checkState(
+          !settingsMap.containsKey(emulator),
+          "Cannot call addEmulatedService twice for " + emulator.toString());
+      this.settingsMap.put(emulator, settings);
+      return this;
+    }
+
+    @NonNull
+    public EmulatorSettings build() {
+      return new EmulatorSettings(settingsMap);
```

Comment:
nit: consider making a copy of the map, otherwise if a reference to the builder is kept around, one can modify the underlying map unintentionally.

Reply:
That is done in the `EmulatorSettings()` constructor with `unmodifiableMap`

Reply:
The problem with `unmodifiableMap` is that it only a read-only **view** into the underlying map, so modifications to the underlying map will be seen by the view.

Reply:
Ah thanks for the clarification, resolved.